File: valistream-windows/valistream/validator/spec_catalog.py
"""Rule-to-spec citation catalog for findings grounded in an external specification."""

from __future__ import annotations

_RFC8216_PREFIX = "RFC8216."

_EXPLICIT: dict[str, str] = {
    "APPLE.codecs": "HLS Authoring §9.1",
    "APPLE.resolution": "HLS Authoring §9.2",
    "APPLE.independent-segments": "HLS Authoring §9.11",
    "APPLE.average-bandwidth": "HLS Authoring §9.14",
    "APPLE.iframe-playlists": "HLS Authoring §6.1",
    "TOOL.continuity.media-sequence": "RFC 8216 §6.2.2",
    "TOOL.continuity.head-removal": "RFC 8216 §6.2.2",
    "TOOL.continuity.segment-stability": "RFC 8216 §6.2.2",
    "TOOL.continuity.discontinuity-inserted": "RFC 8216 §4.3.2.3",
    "TOOL.continuity.discontinuity-sequence": "RFC 8216 §4.3.3.3",
    "TOOL.staleness": "RFC 8216 §6.2.1",
}


def reference(rule_id: str) -> str | None:
    if rule_id.startswith(_RFC8216_PREFIX):
        remainder = rule_id[len(_RFC8216_PREFIX):]
        dash = remainder.find("-")
        section = remainder[:dash] if dash != -1 else remainder
        if not section:
            return None
        return f"RFC 8216 §{section}"
    return _EXPLICIT.get(rule_id)

File: sources/Valistream/Sources/ValistreamCore/Validation/SpecCatalog.swift
//
//  SpecCatalog.swift
//  ValistreamCore
//
//  Created by Volodymyr Akimenko on 17/06/2026.
//

/// Rule-to-spec citation catalog for findings grounded in an external specification.
enum SpecCatalog {
    // MARK: - Internal

    static func reference(forRuleId id: String) -> String? {
        if id.hasPrefix("RFC8216.") {
            let remainder = id.dropFirst("RFC8216.".count)
            let section = remainder.prefix { $0 != "-" }
            guard section.isEmpty == false else { return nil }

            return "RFC 8216 §\(section)"
        }

        return explicitReferences[id]
    }



    // MARK: - Private

    private static let explicitReferences = [
        "APPLE.codecs": "HLS Authoring §9.1",
        "APPLE.resolution": "HLS Authoring §9.2",
        "APPLE.independent-segments": "HLS Authoring §9.11",
        "APPLE.average-bandwidth": "HLS Authoring §9.14",
        "APPLE.iframe-playlists": "HLS Authoring §6.1",
        "TOOL.continuity.media-sequence": "RFC 8216 §6.2.2",
        "TOOL.continuity.head-removal": "RFC 8216 §6.2.2",
        "TOOL.continuity.segment-stability": "RFC 8216 §6.2.2",
        "TOOL.continuity.discontinuity-inserted": "RFC 8216 §4.3.2.3",
        "TOOL.continuity.discontinuity-sequence": "RFC 8216 §4.3.3.3",
        "TOOL.staleness": "RFC 8216 §6.2.1"
    ]
}
